Split `TableModel` into separate classes

hamogu opened the pull request:
## Summary
Split the `TableModel` into two separate models: `FixedTableModel` for a model where the number of bins is fixed to be the same number as the data values and `InterpolatedTableModel1D` where a tabulated x/y table is interpolated (and integrated if needed) to match the data values/bins.

## Details

The old `TableModel` has two separate applications: It can be an x/y table that is interpolated in x as needed, or it can just be a table of y values assuming that the number of values exactly matches the number of data values.
    
This is conceptually different and can be confusing to users (and also in the code). This commit introduces two new classes to split the two different functionalities into separate classes (with a common base class).

Adding those two new classes is followed by:

- deprecating `TableModel`
- Adjusting the models generated by UI functions that load a table

in doing so, this includes the changes in #2376, which change the loading functions in a way that is technically breaking API, but in practice I judge that to be a minimal impact which allows for significantly more consistency in the API (I would just treat the change as a "bug fix"):  align the signature of the `load_table_model` in `sherpa.ui` and `sherpa.astro.ui`. This is technically an API breaking change in case someone relied on the ordering of arguments and passed in the interpolation method as third argument instead of a keyword argument.

I think that's unlikely to happen in practice (most people won't pass that in at all and just use the default), but it's separated into it's own commit in case we want to discuss or reverse that change.

Matching the API does not only make the linters happy, it also simplifies documentation as we could now point to (or generate from in some way) the docs of this method from the docs of the superclass method.

In fact, I'm removing keywords that are no longer used, because they are confusing. Again, I don't think any of them are used in practice, but to keep backwards compatibility, we could just keep those unused keywords in the signature and raise a DeprecationWarning if they are set.

Also, this changes some error messages. It raises the same type of exceptions as before (IoErr) but with a different string message. Again, that's technically breaking the API, but I don't think it changes any workflow in practice.

## Questions

We could keep more of API backwards compatible (a re-ordering of keywords is required, but we can keep the old keywords around (in a differenet order) to blunt the impact).

closes #1648
closes #1620
fixes #1056
closes #2376

hamogu committed:
Remove duplication of the `load_table_model` method between `sherpa.ui` and `sherpa.astro.ui`.

To implement this I align the signature of the `load_table_model` in `sherpa.ui` and `sherpa.astro.ui`.
This is technically an API breaking change in case someone relied on the ordering of arguments and passed in the interpolation method as third argument instead of a keyword argument.

I think that's unlikely to happen in practice (most people won't pass that in at all and just use the default), but it's separated into it's own commit in case we want to discuss or reverse that change.

Matching the API does not only make the linters happy, it also simplifies documentation as we could now point to (or generate form in some way) the docs of this method from the docs of the superclass  method.

In fact, I'm removing keywords that are no longer used, because they are confusing. Agian, I don't think any of them are used in practice, but to keep backwards compatibility, we could just keep those unused keywords in the signature and raise a DeprecationWarning if they are set.

closes #1648

hamogu committed:
New Table models to split functionality of existing TableModel

TableModel really has two separate applications: It can be an x/y table that is interpolated in x as needed, or it can just be a table of y values assuming that the number of values exactly matches the number of data values.

This is conceptually different and can be confusing to users (and also in the code). This commit introduces two new classes to split the two different functionalities into separate classes (with a common base class).

hamogu committed:
Add a deprecation warning for TableModel

This and the previous commited together
closes #1620
fixes #1056

hamogu committed:
Replace use of `TableModel` with `FixedTableModel` or `InterpolatedTableModel1D` throughout Sherpa's codebase.

hamogu commented:
Test failure is unrelated; the failing test probably just needs a slightly larger numerical range to pass on all Python and numpy versions and all OS that we test. I'll address that in a separate PR, but for now, please assume that this is ready to review.

hamogu commented:
rebased onto https://github.com/sherpa/sherpa/pull/2453 as it changes the DS9 download code (and updates versions).

@DougBurke if you could find time to review this, that would be great!

DougBurke commented on sherpa/models/basic.py:
```diff
@@ -2184,6 +2187,397 @@ def calc(self, p, x0, x1=None, *args, **kwargs):
 
 
+
+class TableModelBase(ArithmeticModel, metaclass=ABCMeta):
```
Are we "safe" using a  `metaclass` here? Two points

1. this is in the middle of the class hierarchy (i.e. this is not the base class); is this okay (e.g. the `Model` class is derived from `NoNewBlahBlah` and I thought this would be an issue, but I could be wrong)
2. I have previously seen some surprising behavior after some work where I tried moving to `super()` rather than calling the parent class directly, which has made me wary of this (unfortunately I can't remember the details). In this case it's probably okay because we don't expect there to be many complicated classes that derive from `TableModelBase`

hamogu replied:
I'm pretty sure it's OK. First, the fact that all the old and new tests are passing is a good sign. 
More specifically:

1. We can add meta classes at any level and a class can have multiple metaclasses. That doesn't change the applicability of the other metaclasses, unless it implements methods of the same name. See code example below.
2. In simple inheritance, `super` and directly calling the superclass is interchangeable. Differences arise when you have a diamond structure or something even more complicated because `super` might then lead to a different call order than selecting which superclass to call directly. That's not an issue here. `super()` will call to `ArithmeticModel` because it's the only normal (not meta) class in the inheritance scheme. (It might be possible to screw this up if you derive a class from `TableModelBase` that also has some other child of `ArithmeticModel` as a parent. But that's (a) unlikely to make any sense and (b) we don't know which behavior would be preferred in that case so I'm not worrying about it.) We could replace "super" with "ArthimeticModel" here and it won't make a difference, if you think that reads better.

Example showing that `NoNewBlaBlah` still works:
```Python
>>> from sherpa.models.basic import InterpolatedTableModel1D
>>> tm = InterpolatedTableModel1D()
>>> tm.load([10, 20, 25, 30], [14, 12, 17, 18])
>>> tm.ampl = 10
>>> tm([15, 20, 27])
array([130., 120., 174.])
>>> tm.sergasdf = 4
---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
Cell In[4], [line 1](vscode-notebook-cell:?execution_count=4&line=1)
----> [1](vscode-notebook-cell:?execution_count=4&line=1) tm.sergasdf = 4

File ~/code/sherpa/sherpa/models/model.py:711, in Model.__setattr__(self, name, val)
    708     par.val = val
    709     return
--> [711](https://file+.vscode-resource.vscode-cdn.net/Users/guenther/code/sherpa/debugging/~/code/sherpa/sherpa/models/model.py:711) NoNewAttributesAfterInit.__setattr__(self, name, val)
    713 # If adding a parameter, add the entry to _par_index and
    714 # handle any aliases.
    715 #
    716 if not isinstance(val, Parameter):

File ~/code/sherpa/sherpa/utils/__init__.py:133, in NoNewAttributesAfterInit.__setattr__(self, name, val)
    131 def __setattr__(self, name: str, val: Any) -> None:
    132     if self.__initialized and (not hasattr(self, name)):
--> [133](https://file+.vscode-resource.vscode-cdn.net/Users/guenther/code/sherpa/debugging/~/code/sherpa/sherpa/utils/__init__.py:133)         raise AttributeError(f"'{type(self).__name__}' object has no attribute '{name}'")
    135     if self.__initialized and hasattr(self, name):
    136         cname = callable(getattr(self, name))

AttributeError: 'InterpolatedTableModel1D' object has no attribute 'sergasdf'
```

DougBurke commented on sherpa/models/tests/test_tablemodel.py:
```diff
@@ -31,68 +31,82 @@
     """What is the ndim?
 
     At the moment we can use this with nD data but only really if we do
     not set the "x" argument.
```
```suggestion
```

As the code now kind-of sets ndim the comment is now out-of-date and I'm not sure it's worth adding anything to describe the current behavior.

DougBurke commented on sherpa/models/basic.py:
```diff
@@ -2184,6 +2187,397 @@ def calc(self, p, x0, x1=None, *args, **kwargs):
+        raise ModelErr("filtermismatch", 'table model',
+                       f"data, ({len(self._y)} vs {len(x0)})")
+
+class InterpolatedTableModel1D(TableModelBase, RegriddableModel1D):
```
What benefit do we have from making this part of `RegriddableModel1D`? 

The regrid stuff was integrated into sherpa without an obvious discussion [*] of the table model class, which already had some form of "regrid" capabilities - e.g. the `calc` method below. So I see two possibilities

- can we use the regrid code to avoid some of the code here 
- if not, then I am concerned about having two different ways to get the same [**] thing done, which could be confusing


---

[*] well, I don't remember it, which doesn't actually mean it wasn't discussed

[**] of course, they aren't quite the same thing

hamogu replied:
I don't think we strictly speaking need it. At least our tests pass without inheriting from `RegriddleModel1D`. I put it there because I think that's what is conceptually does - regriddable models promise to be regriddable, which this model is. The `Regriddable_XXX` classes are actually very small and have only one function (`regrid`) in them, so there isn't any code that could be reused here and I'm not concerned about having two ways to do something that is not the same thing.

DougBurke commented on sherpa/astro/ui/utils.py:
```diff
@@ -11224,11 +11234,6 @@ def load_table_model(self, modelname, filename,
            The name of the file containing the data, which should
            contain two columns, which are the x and y values for
            the data, or be an image.
-        method : func
```
[i have many words that could be said about documenting python code in the presence of kwargs, so some "thoughts" may skip through here. This is definitely based on having to dig through matplotlib documentation hunting down arguments]

In this case the use of `kwargs` lets us drop some repeated code. However, for the user it is, I claim, a demonstrably poorer experience, as the have no idea about the `method` argument or its default value. We don't even have an example showing it (not that having an example would absolve us, I think).

I'm not sure of the best thing to do here. Matlpotlib can (in some cases at least) point to other documentation when describing the `kwargs` argument, but that's not always possible, or makes sense here. One option is to leave in just the documentation. Another would be to not remove the `method=...` argument above. Or leave as is.

hamogu replied:
The DOC has a longstanding "DOC-TODO: how to describe the supported args/kwargs (not just for this function)?" comment above its signature indicating that this problem isn't new. What's new is that this function is now fully redundant. It doesn't do anything but calling the super class and is only here so we get a chance to write down a docstring. I'll expand that to be more useful.

hamogu committed:
Improve documentation for load_table_model

hamogu committed:
Minor comments and doc fixes

DougBurke commented on sherpa/astro/ui/tests/test_astro_ui.py:
```diff
@@ -22,6 +22,7 @@
 import os
 import re
 import logging
+from tempfile import NamedTemporaryFile
```
An alternative to using `NamedTemporaryFile` is the [tmp_path](https://docs.pytest.org/en/latest/how-to/tmp_path.html) fixture but I'm not sure it's worth changing here.

hamogu replied:
They are functionally identical as far as I can tell and I don't care which one to use. If you don't have a preference, let's just leave it as is.
(Maybe "temp_path" is from a time when the "NameTemporaryFile" wasn't in the standard library yet or it offers additional logging for debugging or something like that?)

DougBurke commented on sherpa/models/basic.py:
```diff
@@ -2017,6 +2021,10 @@ def __init__(self, name='tablemodel'):
         self._method = linear_interp
         self.ampl = Parameter(name, 'ampl', 1)
         ArithmeticModel.__init__(self, name, (self.ampl,))
+        warn("TableModel is deprecated and will be removed in a future version. "
```
We should add a docstirng at line 1888 ish along the lines of (with left padding added)

```
... deprecated:: 4.19.0
    Use the FixedTableModel or InterpolatedTableModel1D classes instead.
```

DougBurke commented:
We need somehting like

```
% git diff .
diff --git a/docs/model_classes/basic.rst b/docs/model_classes/basic.rst
index c6c33f12..479acb78 100644
--- a/docs/model_classes/basic.rst
+++ b/docs/model_classes/basic.rst
@@ -19,8 +19,10 @@ The sherpa.models.basic module
       Erfc
       Exp
       Exp10
+      FixedTableModel
       Gauss1D
       Integrate1D
+      InterpolatedTableModel1D
       Log
       Log10
       LogParabola
@@ -34,6 +36,7 @@ The sherpa.models.basic module
       StepHi1D
       StepLo1D
       TableModel
+      TableModelBase
       Tan
       UserModel
 
@@ -49,5 +52,5 @@ The sherpa.models.basic module
 Class Inheritance Diagram
 =========================
 
-.. inheritance-diagram:: Box1D Const1D Scale1D Cos Delta1D Erf Erfc Exp Exp10 Gauss1D Integrate1D Log Log10 LogParabola NormGauss1D Poisson Polynom1D PowLaw1D Sin Sqrt StepHi1D StepLo1D TableModel Tan UserModel Box2D Const2D Scale2D Delta2D Gauss2D SigmaGauss2D NormGauss2D Polynom2D
+.. inheritance-diagram:: Box1D Const1D Scale1D Cos Delta1D Erf Erfc Exp Exp10 FixedTableModel Gauss1D Integrate1D InterpolatedTableModel1D Log Log10 LogParabola NormGauss1D Poisson Polynom1D PowLaw1D Sin Sqrt StepHi1D StepLo1D TableModel TableModelBase Tan UserModel Box2D Const2D Scale2D Delta2D Gauss2D SigmaGauss2D NormGauss2D Polynom2D
    :parts: 1
```

to add the new classes to the Read-The-Docs documentation (although when I just tried it sphinx warned me about

```
/Users/burked/code/sherpa/sherpa-1/sherpa/models/basic.py:docstring of sherpa.models.basic.InterpolatedTableModel1D.method:1: WARNING: duplicate object description of sherpa.models.basic.InterpolatedTableModel1D.method, other instance in model_classes/api/sherpa.models.basic.InterpolatedTableModel1D, use :no-index: for one of the
```

which I have no energy to investigate just now.

There is the question of whether the `TableModelBase` class should be included, as it isn't an exported symbol, but I err on the case of including it.

DougBurke commented on sherpa/astro/ui/utils.py:
```diff
@@ -11245,14 +11262,63 @@ def load_table_model(self, modelname, filename,
-        are: `linear_interp`, `nearest_interp`, `neville`, and
-        `neville2d`.
+
+        * Keywords for ASCII tables*
```
```suggestion
        * Keywords for ASCII tables
```

or did you mean to have the asterix here? It looks a bit odd in the sphinx doc build: e.g.

<img width="827" height="1046" alt="Image" src="https://github.com/user-attachments/assets/d045d3d7-227a-4a4c-ba97-62f3a277258e" />

hamogu replied:
No, I only wanted the asterisk in the front (which makes it a bulleted list). I experimented with headings (which I couldn't get sphinx to render at the right level, presumably because it's inside a numpy-style docstring), bold font, numbered and bulleted lists. None is quite perfect for this, but I settled on the list. The extra star must have slipped though from one of the other experiments.

DougBurke replied:
I feel your pain for the "trying to work out how to get the documentation look nice" attempt...

hamogu commented:
> /Users/burked/code/sherpa/sherpa-1/sherpa/models/basic.py:docstring of sherpa.models.basic.InterpolatedTableModel1D.method:1: WARNING: duplicate object description of sherpa.models.basic.InterpolatedTableModel1D.method, other instance in model_classes/api/sherpa.models.basic.InterpolatedTableModel1D, use :no-index: for one of the

This seems to be a known limitation of autodoc that gets confused when the name of a property and the name of a parameter to `__init__` for a class are identical. While I have some general thoughts about autodoc that's beyond the scope if this PR. The parameter is new. In the old table class the only way to change it was to set the property later. I don't like that pattern in general, because it's annoying to make a model and then change it if you want a different default. But since it's new, we don't have to worry about backwards compatibility and I'll just change the name of the parameter.

hamogu committed:
Clear cache also when integration setting are changed.
Do not expose those setting in __init__ as a parameter. That is mostly because the sphinx autodoc extension raises a warning if the same mane is both a property and a parameter in __init__. Workarounds to change the default settings might influence other docstrings, too, and are of the scope for this PR.

This PR also include minor typo fixes.

hamogu committed:
Fix sphinx formatting

DougBurke commented on sherpa/models/basic.py:
```diff
@@ -2472,21 +2477,24 @@ class InterpolatedTableModel1D(TableModelBase):
+                 x: Sequence | None=None, y: Sequence | None=None,
+                 method: Callable = linear_interp,
+                 integrate_method: Callable = integrate_tabulated_function,
+                 integrate_kwargs: dict = {}):
```
Should this default to `None` - which we then replace with the empty `dict` ? There is the whole "do not use a mutable argument" that `pylint` complains about (but I note `ruff` does not):

https://pylint.readthedocs.io/en/stable/user_guide/messages/warning/dangerous-default-value.html

hamogu replied:
That's a good point. Mutable default argument make for hard-to-debug error messages. Will change.

hamogu committed:
Avoid a mutable default argument

DougBurke commented:
Note that the RTD failure is fixed in the main branch (#2473) and it is completely unrelated to this PR

DougBurke commented on sherpa/astro/ui/utils.py:
```diff
@@ -1368,7 +1370,7 @@ def unpack_arrays(self, *args):
             return sherpa.astro.io.read_arrays(*args)
         except NotImplementedError:
             # if the astro backend is not set, fall back on io module version.
```
This comment is not out of date (as the fall back is no longer the `sherpa.io.read_arrays` call). We could fix it but I think it's obvious from the context so lnie 1372 can just be removed.

```suggestion
```

DougBurke commented on sherpa/models/basic.py:
```diff
@@ -18,12 +18,15 @@
-from sherpa.utils import bool_cast, interpolate, linear_interp, \
-    sao_fcmp
+from sherpa.utils import bool_cast, integrate_tabulated_function, interpolate, linear_interp, \
+    integrate, sao_fcmp
```
```suggestion
    sao_fcmp
```

I don't think we use the `integrate` symbol, but I could be mistaken.

hamogu replied:
I use it in line 2619 in the same file in `IntegratedTableModel1D` like this:
```Python
    @modelCacher1d
    def _calc_integrate(self, x0, x1):
        return integrate(x0, x1, self._x, self._y,
                         interpolate_func=self.method,
                         integrate_func=self.integrate_method,
                         **self.integrate_kwargs)
```

DougBurke commented on sherpa/utils/__init__.py:
```diff
@@ -34,7 +34,7 @@
 import sys
 from types import FunctionType, MethodType
-from typing import Any, Generic, TypeVar, overload
+from typing import Any, Dict, Generic, TypeVar, overload
```
```suggestion
from typing import Any, Generic, TypeVar, overload
```

We don't need this anymore.

DougBurke commented on sherpa/utils/__init__.py:
```diff
@@ -2344,6 +2345,121 @@ def interpolate(xout, xin, yin, function=linear_interp):
+
+def integrate(x0: np.ndarray, x1: np.ndarray, xin: np.ndarray, yin: np.ndarray,
+              interpolate_func: Callable[[np.ndarray, np.ndarray, np.ndarray], np.ndarray] = linear_interp,
+              integrate_func: Callable[[np.ndarray, np.ndarray, np.ndarray, np.ndarray, Callable, Dict], np.ndarray] = integrate_tabulated_function,
```
```suggestion
              integrate_func: Callable[[np.ndarray, np.ndarray, np.ndarray, np.ndarray, Callable, dict | None], np.ndarray] = integrate_tabulated_function,
```

As mentioned above, we don't need `Dict`. I'm pretty sure (but have not checked as much as I should have) that the "mutable fix" you made today means that the "`| None`" change is now also needed.

DougBurke commented on sherpa/ui/utils.py:
```diff
@@ -7950,7 +7943,7 @@ def func2d(pars, x0, x1, x0hi=None, x1hi=None):
         if filename is not None:
             _, usermodel._y = self._read_user_model(filename, ncols, colkeys,
-                                                    dstype, sep, comment)
+                                                    sep, comment)
```
I think this is the largest conceptual issue I have here. As far as I can see we now longer let the user pass through a `Data2D` or `Data2Dint` case.

I am 99.99% sure we have no tests of this functionality (definitely not in the Sherpa test suite, and probably not in CIAO itself) so it could well be that the code never worked sensibly.

hamogu replied:
Yes, this is a deliberate decision (which iI intended to put into its own commit so we could reverse it but it turned out that there is no reasonable way to carry that through with the other changes in this PR). Also see discussion on #2376.

The short summary is that, as fas as I can tell, that never really worked in general, was never tested, and never advertised. The basic problem is that there that all those formats look too similar, i.e. a 2D (unwrapped as above) looks exactly like a 1DInt (x_lo, x_hi, y) in the file format. 

\<rant\>
There is a web (literally, I drew it out on the piece of paper) of private and partially hidden  functions like the `_read_user_model` that is called from the public `load_user_model` and many of those involve heuristics and guessing to identify what the format of the file should be, even when a specific data class is passed. There are just too many options. Even a simple Data1D can have anywhere between two (x, y) and four columns (x, y, stat_error, sys_error). How do you distinguish between that and Data2DInt with three to five columns or any of the datasets in between in terms of complexity (Data1DInt, Data2D)? Now, you might say "that's exactly why we had that keyword", but in fact the code is not written that way. Even with that keyword passed in, the underlying code is still making guesses about how many columns there are and what they mean. 
`_read_user_model` used to have multiple try..except blocks nested to encompass that, which I've changed to a simpler loop, but the concept is the same (which is why all old and new tests pass).
It's like a design that contains only edge cases; it's what an LLM would have written if it had been around 10 years ago and the instructions were given by monkeys who also don't know what they want. Ok, I'm joking for the last one: The truth is that the programmer of that code probably wanted to make it work out-of-the-box for almost any format the user throws at it.
We want it to "just work" (TM)! 
And then some other programmer tried the same for the `astro.ui.utils`, with slightly different assumptions. So we have a second version there, that's almost, but not quite giving the same answers leading to subtle inconsistencies that have driven me crazy in debugging before (which resulting in #2376).
But if you make it work so flexibly for almost anything that you throw at it, then it can't be specific at the same time. The vast majority of use cases (I wager about 101 % of them) will do this for some type of 1D data (or  data we thing of as 2SD in our head, but that's unwrapped in Sherpa so that it's Data1D or Data1DInt). I've looked long and hard when I did those changes, because I know about the pain of backwards braking changes, and I cannot see a way to make a sensible use of the `dstype` keyword in the structure that we have.
(You could write essentially separate functions for 1D and 2D but unless we have a use case for that, we shouldn't add that complexity. And even when we have a use case, we should suggest a more direct way to do that.)

We did something similar with astropy, where we have a function to parse almost any ASCII table. That sounds great until you run into edge cases like "is `observation start`" one column or two, separated by a space? What if I want it to be two columns? What about comment characters? What is every line is commented? Try a different format, or assume that user wants to read a table with no data lines? Now, we try not to touch that code any more and if we add new formats, users have to pass them in explicitly. 
It's kind of the same here: We also read ASCII tables (and fits), and the format (in particular in the ASCII case) is simply non-unique, so, mathematically speaking, there can be no way to get all reads right. But the vast majority of use cases (and all tests) are for 1D reads, so we can reduce a lot of that ambiguity by assuming it's always 1D and for the 2D cases ask users to write like 2 lines of python themselves because they know the formats of their input data and don't have to guess:
```
intab = np.loadtxt(filename, parameters)
mydata =Data2DInt(intab[0,:], intab[1, :], .... whatever ...)
--- or ---
ui.load_arrays('mydataset', intab[0,:], intab[1, :], .... whatever ..., Data2DInt)
```

Trying to be 100% backwards compatible here would get us another rat's nest of ill defined behavior, that's hard to test, creates new edge cases, and will probably never be used in the wild. (Because people who built their own Data2DInt models are likely to be users familiar with Python who would rather load the data themselves than spending time to discover exactly how the API works here. For non-specialist cases, we have, e.g. "load_psf"; those people wouldn't encounter this function I changed here.)

\<end rant\>

DougBurke commented:
So, I've recently been looking at out C API and integration, and I wonder whether it's relevant here. However, it is completely undocumented and is all about passing around C routines between modules and is used by `sherpa.models.basic.Integrator1D`. 

So there's possibility some clean up that could be done later to try and unify some of this "integrate" code.

hamogu commented:
I agree that some parts of the C API could be redone or replaced by e.g. scipy (which is already on conda-forge so we don't have to built it), but that's definitely out-of-scope here.

hamogu commented:
And by "redone" I don't mean "rewriting the C-code", but "expose it at the Python level and/or improve the way it's visible to users".
However, unless we have a specific use case, that might not be the bast use of our time. It works now where we need it, so at least I won't touch it unless I have to.

hamogu commented:
Rebased and pushed to pick up the changes on main to make CI pass. Also including the small fixes suggested by @DougBurke

DougBurke commented on sherpa/utils/__init__.py:
```diff
@@ -2344,6 +2345,121 @@ def interpolate(xout, xin, yin, function=linear_interp):
+
+def integrate(x0: np.ndarray, x1: np.ndarray, xin: np.ndarray, yin: np.ndarray,
+              interpolate_func: Callable[[np.ndarray, np.ndarray, np.ndarray], np.ndarray] = linear_interp,
+              integrate_func: Callable[[np.ndarray, np.ndarray, np.ndarray, np.ndarray, Callable, dict | None], np.ndarray] = integrate_tabulated_function,
```
Technically the signature for `integrate_func` is incorrect since `Callable` can't handle keyword arguments and in line 2458 we call it with an explicit keyword argument and then pass it `**kwargs`. However, it expresses the intent and we can look at improving it at some point in the future when  the `Callable` / `Protocol` support has stabilized for the python versions we support.

For anyone following along at home, I noticed this from:

```
% pyright sherpa/utils/__int__.py
...
  /lagado.real/sherpa/sherpa-main/sherpa/utils/__init__.py:2459:44 - error: Expected 2 more positional arguments (reportCallIssue)
...
```

DougBurke replied:
So this is not a comment on this PR per se - just something that we can improve down the line.

DougBurke commented:
This commit will require some changes to https://github.com/cxcsds/docstring_to_ahelp but that sort of thing is expected and not a blocker.

DougBurke reviewed (approved):
Sorry for the delay in reviewing.

wmclaugh commented on sherpa/utils/__init__.py:
```diff
@@ -2344,6 +2345,121 @@ def interpolate(xout, xin, yin, function=linear_interp):
+                                 *args, **kwargs):
+    '''Integrate y_points between x0 and x1
+
+    This function is meant to be used for models that linarly interpolate
```
```suggestion
    This function is meant to be used for models that linearly interpolate
```

DougBurke replied:
@hamogu is travelling so I'll just accept this change for him.

DougBurke committed:
Update sherpa/utils/__init__.py

Co-authored-by: wmclaugh <wmclaugh@cfa.harvard.edu>
